src/main/java/com/pingcap/tikv/operation/ChunkIterator.java

```diff
@@ -23,59 +23,78 @@
+public abstract class ChunkIterator<T> implements Iterator<T> {
 
-    ChunkIterator(List<Chunk> chunks) {
-      // Read and then advance semantics
-      this.chunks = chunks;
-      chunkIndex = 0;
-      metaIndex = 0;
-      bufOffset = 0;
-      if (chunks.size() == 0
-          || chunks.get(0).getRowsMetaCount() == 0
-          || chunks.get(0).getRowsData().size() == 0) {
-        eof = true;
-      }
-    }
-
-    @Override
-    public boolean hasNext() {
-      return !eof;
-    }
+  private final List<Chunk> chunks;
+  protected int chunkIndex;
+  protected int metaIndex;
+  protected int bufOffset;
+  protected boolean eof;
 
-    private void advance() {
-      if (eof) return;
-      Chunk c = chunks.get(chunkIndex);
-      bufOffset += c.getRowsMeta(metaIndex++).getLength();
-      if (metaIndex >= c.getRowsMetaCount()) {
-        // seek for next non-empty chunk
-        do {
-          chunkIndex += 1;
-        } while (chunkIndex < chunks.size() && chunks.get(chunkIndex).getRowsMetaCount() == 0);
-        if (chunkIndex >= chunks.size()) {
-          eof = true;
-          return;
+  public static ChunkIterator<ByteString> getRawBytesChunkIterator(List<Chunk> chunks) {
+    return new ChunkIterator<ByteString>(chunks) {
+      @Override
+      public ByteString next() {
+        Chunk c = chunks.get(chunkIndex);
+        long endOffset = c.getRowsMeta(metaIndex).getLength() + bufOffset;
+        if (endOffset > Integer.MAX_VALUE) {
+          throw new TiClientInternalException("Offset exceeded MAX_INT.");
         }
-        metaIndex = 0;
-        bufOffset = 0;
+
+        ByteString result = c.getRowsData().substring(bufOffset, (int) endOffset);
+        advance();
+        return result;
       }
+    };
+  }
+
+  public static ChunkIterator<Long> getHandleChunkIterator(List<Chunk> chunks) {
+    return new ChunkIterator<Long>(chunks) {
+      @Override
+      public Long next() {
+        Chunk c = chunks.get(chunkIndex);
+        long result = c.getRowsMeta(metaIndex).getHandle();
+        advance();
+        return result;
+      }
+    };
+  }
+
+  protected ChunkIterator(List<Chunk> chunks) {
+    // Read and then advance semantics
+    this.chunks = chunks;
+    this.chunkIndex = 0;
+    this.metaIndex = 0;
+    this.bufOffset = 0;
+    if (chunks.size() == 0
+        || chunks.get(0).getRowsMetaCount() == 0
+        || chunks.get(0).getRowsData().size() == 0) {
+      eof = true;
     }
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !eof;
+  }
 
-    @Override
-    public ByteString next() {
-      Chunk c = chunks.get(chunkIndex);
-      long endOffset = c.getRowsMeta(metaIndex).getLength() + bufOffset;
-      if (endOffset > Integer.MAX_VALUE) {
-        throw new TiClientInternalException("Offset exceeded MAX_INT.");
+  protected void advance() {
+    if (eof) {
+      return;
+    }
+    Chunk c = chunks.get(chunkIndex);
+    bufOffset += c.getRowsMeta(metaIndex++).getLength();
```

Comment:
what does `bufOffset` mean here since it is not used

Reply:
this is a state of iterator itself. it is used in next to decide next part inside chunk.